src/Tactic/Premises.agda

```diff
@@ -0,0 +1,294 @@
+
+{-
+** Extracting the hypotheses of an STS rule.
+
+Constructors should be formulated in the STS syntax for inference rules, i.e.
+```
+  rule : ∀ <IMPLICIT_ARGUMENTS + LET_EXPRESSIONS>
+    ∙ <HYPOTHESIS#1>
+    ∙ <HYPOTHESIS#2>
+    ⋮
+    ∙ <HYPOTHESIS#n-1>
+    ∙ <HYPOTHESIS#n>
+      ────────────────────────────────
+      <CONCLUSION>
+```
+This meta-program then generates the type of the (decidable) hypotheses of the rule,
+as well as a proof that it is in fact decidable, i.e.
+```
+rule-premises : ∀ ⋯ → Set
+rule-premises ⋯ =
+  ∙ <HYPOTHESIS#1>
+  ∙ <HYPOTHESIS#3>
+  ⋮
+  ∙ <HYPOTHESIS#n-1>
+
+rule-premises? : ∀ ⋯ → Dec (rule-premises ⋯)
+rule-premises? = it
+```
+where ⋯ stands for the subset of implicit arguments that are relevant for the
+selected subset of decidable hypotheses.
+
+To be precise, we generate both definitions above under a single name, i.e.
+```
+unquoteDecl rule-premises = genPremises rule-premises (quote rule)
+-- rule-premises        : ∀ ⋯ → ∃ Dec
+-- rule-premises .proj₁ : ∀ ⋯ → Set
+-- rule-premises .proj₂ : ∀ ⋯ → Dec (rule-premises .proj₁ ⋯)
+```
+-}
+genPremises : Name → Name → TC ⊤
+genPremises f n = do
+  print $ "*** Generating premises for constructor:" <+> show n <+> "***"
+  ty ← reduceRuleSyntax =<< getType n                  -- (1)
+  let is , hs = breakImplicits $ viewTy ty .proj₁      -- (2)
+  hs′ ← map unbundleHypotheses                         -- (3)
+      $ extendContextTel (absTelescope is)
+      $ mapM reduceRuleSyntax hs
+  dhs ← forM hs′ $ λ h → (_, h) <$> isDecidable? is h  -- (4)
+  let dhs′ = proj₂ <$> filter ((_≟ true) ∘ proj₁) dhs
+  is , premise ← removeUnusedImplicits is -- (5)
+               $ bundleHypotheses dhs′    -- (6)
+  let premiseTy  = tyView (is , quote ∃Dec ∙)
+      premiseTel = absTelescope is
+      premisePs  = map (λ (i , x) → arg (argInfo $ unAbs x) (` toℕ i))
+                       (uncurry L.zip $ map₁ reverse $ unzip $ enumerate is)
+      ∃premise?  = quote _,_ ◆⟦ premise ∣ quote it ∙ ⟧
+      cl         = clause premiseTel premisePs ∃premise?
+  print $ "```\n" <> show f <+> ":" <+> show premiseTy <> "\n"
+       -- <> show f <+> "⋯ =" <+> show premise <> " , it\n```\n"
+       -- TOO EXPENSIVE FOR LARGE RULES (even without `-v` flag)
+  declareDef (vArg f) premiseTy
+  defineFun f [ cl ]
+  where
+  -- (1) unfold syntactic sugar from Interface.STS
+  reduceRuleSyntax : Type → TC Type
+  reduceRuleSyntax ty = do
+    let whitelist = Interface.STS._∙_
+                 `∷ Interface.STS.∙_
+                 `∷ Interface.STS._────────────────────────────────_
+                 `∷ Interface.STS.────────────────────────────────_
+                 `∷ Interface.STS._───────────────────────────────────────_
+                 `∷ Interface.STS.───────────────────────────────────────_
+                 -- ^ include all syntactic sugar for inference rules here
+                 `∷ []
+    onlyReduceDefs whitelist $ go GAS_LIMIT ty
```

Comment:
`onlyReduceDefs whitelist $ normalise ty` works too, is there a reason you didn't just do that?

Reply:
That was my first attempt, but complicated rules like UTXOW take too much time/RAM to completely normalise, so it is necessary to have a limit here.

Reply:
Hmm, printing must have intervened when I was noticing the slow behaviour, so great! that simplified the code a lot.